### PR TITLE
cdc: filter out the event cause by cdc write

### PR DESCRIPTION
**Pull request by xiongjiwei**
Signed-off-by: xiongjiwei <xiongjiwei1996@outlook.com>

### What is changed and how it works?

Issue Number: ref https://github.com/tikv/tikv/issues/13779

What's Changed:

```commit-message
filter out the event cause by cdc write
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/cdc/src/delegate.rs [hunk 1]
```diff
@@ -620,6 +631,13 @@ impl Delegate {
         if entries.is_empty() {
             return Ok(());
         }
+        let user_change_entries: Vec<EventRow> = entries
```
Comment (overvenus) [inline components/cdc/src/delegate.rs:634]: Please lazily split entries.

File: components/cdc/src/delegate.rs [hunk 2]
```diff
@@ -926,6 +963,7 @@ fn decode_write(
     row.commit_ts = commit_ts;
     row.key = key.truncate_ts().unwrap().into_raw().unwrap();
     row.op_type = op_type as _;
+    row.txn_source = write.txn_source as u64;
```
Comment (overvenus) [inline components/cdc/src/delegate.rs:966]: Please add some comments.

File: components/cdc/src/delegate.rs [hunk 3]
```diff
@@ -952,6 +990,7 @@ fn decode_lock(key: Vec<u8>, lock: Lock, row: &mut EventRow, has_value: &mut boo
     row.start_ts = lock.ts.into_inner();
     row.key = key.into_raw().unwrap();
     row.op_type = op_type as _;
+    row.txn_source = lock.txn_source as u64;
```
Comment (overvenus) [inline components/cdc/src/delegate.rs:993]: Please add some comments.

File: components/cdc/src/delegate.rs [hunk 4]
```diff
@@ -527,6 +535,9 @@ impl Delegate {
                     row_size = 0;
                 }
             }
+            if row.txn_source != 0 && filter_loop {
+                continue;
```
Comment (overvenus) [inline components/cdc/src/delegate.rs:539]: Please add some comments.

**Review by overvenus — CHANGES_REQUESTED** [review]
Please add tests:
* Test rollback events can be filtered correctly.
* An e2e test in cdc/tests/integrations where we can send an request with filter_loop sets to true.

File: components/cdc/src/delegate.rs [hunk 5]
```diff
@@ -620,6 +632,53 @@ impl Delegate {
...
+            if ds.filter_loop {
+                need_filter = true;
+                break;
+            }
+        }
+
+        // collect the change event cause by user write, which is `txn_source` = 0.
+        // for changefeed which only need the user write, send the `filtered`, or else,
+        // send them all.
+        let (filtered, empty) = if need_filter {
```
Comment (overvenus) [inline components/cdc/src/delegate.rs:654]: What about `Option<Event>` instead of `(filtered, empty)`?

Comment (xiongjiwei) [reply components/cdc/src/delegate.rs]: done, PTAL

File: components/cdc/src/delegate.rs [hunk 6]
```diff
@@ -630,14 +684,23 @@ impl Delegate {
...
             if !downstream.state.load().ready_for_change_events() || downstream.kv_api != kv_api {
                 return Ok(());
             }
-            let event = change_data_event.clone();
+            if downstream.filter_loop && filtered.is_none() {
+                return Ok(());
+            }
+
+            let event = if downstream.filter_loop {
+                filtered.clone().unwrap()
```
[the remark this answers is not in the document]
Comment (xiongjiwei) [reply components/cdc/src/delegate.rs:700]: if code runs here, `downstream.filter_loop` must be true and L695 can ensure `filtered` is not none

**Review by hicqu — APPROVED** [review]

**Review by overvenus — APPROVED** [review]

**Pull request merged**